hikari/impl/special_endpoints.py

```diff
@@ -880,3 +884,108 @@ def build(self, entity_factory: entity_factory_.EntityFactory, /) -> data_bindin
+                raise ValueError("custom_id cannot be specified for a LINK style button")
+
+        else:
+            if self._url is not undefined.UNDEFINED:
+                raise ValueError("url cannot be specified for a non-LINK style button")
+
+            if self._custom_id is undefined.UNDEFINED:
+                raise ValueError("custom_id must be specified for a non-LINK style button")
+
+        if not undefined.any_undefined(self._emoji_id, self._emoji_name):
+            raise ValueError("Only one of emoji_id or emoji_name may be provided")
+
+
+def _build_emoji(
+    emoji: typing.Union[snowflakes.Snowflakeish, emojis.Emojiish, undefined.UndefinedType] = undefined.UNDEFINED
+) -> typing.Tuple[undefined.UndefinedOr[str], undefined.UndefinedOr[str]]:
+    """Build an emoji into the format accepted in buttons.
+
+    Parameters
+    ----------
+    emoji : typing.Union[hikari.snowflakes.Snowflakeish, hikari.emojis.Emojiish, hikari.undefined.UndefinedType]
+        The ID, object or raw string of an emoji to set on a component.
+
+    Returns
+    -------
+    typing.Union[hikari.undefined.UndefinedOr[builtins.str], hikari.undefined.UndefinedOr[builtins.str]]
+        A union of the custom emoji's id if defined (index 0) or the unicode
+        emoji's string representation (index 1).
+    """
+    # Since these builder classes may be re-used, this method should be called when the builder is being constructed.
+    if emoji is not undefined.UNDEFINED:
+        if isinstance(emoji, (int, emojis.CustomEmoji)):
+            return str(int(emoji)), undefined.UNDEFINED
+
+        return undefined.UNDEFINED, str(emoji)
+
+    return undefined.UNDEFINED, undefined.UNDEFINED
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class ActionRowBuilder(special_endpoints.ActionRowBuilder):
+    """Standard implementation of `hikari.api.special_endpoints.ActionRowBuilder`."""
+
+    _components: typing.List[special_endpoints.ComponentBuilder] = attr.field(factory=list, init=False)
+
+    def add_button(
+        self: _ActionRowBuilderT,
+        style: typing.Union[int, components.ButtonStyle],
+        *,
+        label: undefined.UndefinedOr[str] = undefined.UNDEFINED,
+        emoji: typing.Union[snowflakes.Snowflakeish, emojis.Emojiish, undefined.UndefinedType] = undefined.UNDEFINED,
+        custom_id: undefined.UndefinedOr[str] = undefined.UNDEFINED,
+        url: undefined.UndefinedOr[str] = undefined.UNDEFINED,
+        disabled: bool = False,
+    ) -> _ActionRowBuilderT:
+        self._components.append(
+            _ButtonBuilder(
+                *_build_emoji(emoji), style=style, label=label, custom_id=custom_id, url=url, is_disabled=disabled
+            )
+        )
+
+        return self
+
+    def build(self) -> data_binding.JSONObject:
+        return {
+            "type": components.ComponentType.ACTION_ROW,
+            "components": [component.build() for component in self._components],
+        }
```

Comment:
Are these necessary or can they follow a creation pattern like embed does?

Note, I completely missed to ask the same for the other two.

Reply:
These builders don't match up with their relevant API representations and merging the two would infer that the received objects are mutable when they aren't intended to be and make their typing worse/less concise.

Reply:
As far as I can see, they seem to be the same but with some defaults, which can be implemented into the objects

Reply:
id is optional on command builder unlike Command where it is not
Button builder only takes a custom emoji ID whereas Button will have a proper custom emoji object (id and name and whether it's animated)

Reply:
This can be managed by the deserializing and serializing of the model

Reply:
So you want to make Command.id optional?